### ISSUE TITLE
PyLintBear: column number in result violates convention

### ISSUE THREAD

**Issue by li-boxuan** [coala/coala-bears#2393]
According to [coala API doc](https://api.coala.io/en/latest/coalib.results.html#module-coalib.results.Result), column number should be one-based. PyLintBear violates this rule.

How to reproduce:
1. Create a new python file, put a single space in it.
2. Run PyLintBear with --json flag. The sample result is as follows:

```json
{
  "results": {
    "cli": [
      {
        "additional_info": "",
        "affected_code": [
          {
            "end": {
              **"column": 0**,
              "file": "/Users/liboxuan/coala/test.py",
              "line": 1
            },
            "file": "/Users/liboxuan/coala/test.py",
            "start": {
              **"column": 0**,
              "file": "/Users/liboxuan/coala/test.py",
              "line": 1
            }
          }
        ],
        "aspect": "NoneType",
        "confidence": 100,
        "debug_msg": "",
        "diffs": null,
        "id": 289554091275372954659836166155133973976,
        "message": "C0303 - Trailing whitespace",
        "message_arguments": {},
        "message_base": "C0303 - Trailing whitespace",
        "origin": "PyLintBear (C0303)",
        "severity": 0
      }
    ]
  }
}
```
Note that column number should never be 0 since it is one-based.

**Comment by li-boxuan** [comment]
IMO, in this case, column number can be None, but should not be 0.

**Comment by Makman2** [comment]
Why `None`? `None` means the whole line.

**Comment by li-boxuan** [comment]
@Makman2 My bad, I meant to say column number can be None (according to API doc), but cannot be 0. In this specific case, it should not be None but the exact position.

Another issue here: Seems that PyLintBear always reports col number 0.

**Comment by li-boxuan** [comment]
@Makman2 This issue can be solved by transforming the zero-based convention into the one-based convention. I'm willing to solve this.

**Comment by jayvdb** [comment]
@li-boxuan , can you find an example which isnt about something in the middle of the line.  a trailing whitespace could be seen as an error for the whole line.

If pylint is normally 1-based column convention, and there is a few bugs in pylint implementation of their convention, these should be fixed upstream.

If pylint is actually 0-based column convention, then indeed we need to map their column numbers to our column number convention.

**Comment by li-boxuan** [comment]
@jayvdb Although I don't think their documentation mentions their column convention, their tutorial implies that they use 0-based column convention. See the example on https://pylint.readthedocs.io/en/latest/tutorial.html

To illustrate that more clearly, I wrote a sample as follows:
```python
shift = 1
if True:
    choice = 1
```

The report generated by ```pylint``` is as follows:
```
************* Module test
C:  1, 0: Missing module docstring (missing-docstring)
C:  1, 0: Constant name "shift" doesn't conform to UPPER_CASE naming style (invalid-name)
W:  2, 0: Using a conditional statement with a constant value (using-constant-test)
C:  3, 4: Constant name "choice" doesn't conform to UPPER_CASE naming style (invalid-name)
```
It can be seen that "shift" is reported at column 0 while "choice" is reported at column 4.

**Comment by Makman2** [comment]
What about having another parameter for `@linter` that controls this conversion? I can imagine that this is quite common that tools are sometimes zero-based and sometimes one-based with lines and columns.

**Comment by Makman2** [comment]
@li-boxuan Then you don't have to use `process_output` ;)

**Comment by Makman2** [comment]
Filed at https://github.com/coala/coala/issues/5346

**Comment by li-boxuan** [comment]
Should this issue be blocked until that issue is closed?

**Comment by li-boxuan** [comment]
That's true they use 0 if they don't have proper column offsets (https://github.com/PyCQA/pylint/blob/master/pylint/utils.py#L465). 

I'm not 100% sure column offset is zero-based.

I cannot find any strong evidence (no documentation for that). 

Some weak evidence:

1. This issue https://github.com/PyCQA/pylint/issues/104 explains that they use zero-based convention, but it may or may not be out-of-date.

2. As this comment, https://github.com/coala/coala-bears/issues/2393#issuecomment-379483092 illustrates, 
`Constant name "shift" doesn't conform to UPPER_CASE naming style` and `Constant name "choice" doesn't conform to UPPER_CASE naming style` are the same error. The previous one reports column offset 0 and the latter one reports column offset 4.

I would try to dig into upstream of `pylint`, namely `astroid` to verify that.

**Comment by li-boxuan** [comment]
After some investigation, I find out [pylint](https://github.com/PyCQA/pylint) has AST checkers, token checkers, etc. The AST checker makes use of [astroid](https://github.com/PyCQA/astroid) to get `node` info, which actually comes from `ast` module. `ast` uses 0-based column offset ([ast documentation](https://docs.python.org/3/library/ast.html#ast.AST.col_offset) doesn't explicitly say that but the following code verifies that).

```python
>>> import ast
>>> code = ast.parse("print('Hello world!')")
>>> code.body[0].col_offset
0
```

I think we can conclude that `pylint` uses 0-based column offset.

**Comment by jayvdb** [comment]
can you test that again using astroid.  it is not just a light wrappper around ast.

what offset do the token checkers use?  These are the ones I worry are 1 based.

**Comment by li-boxuan** [comment]
Test using astroid (Version: 1.6.3, up-to-date release version):
```python
>>> import astroid
>>> node = astroid.extract_node("print('Hello world!')")
>>> node.__dict__
{'lineno': 1, 'col_offset': 0, 'parent': <Expr l.1 at 0x10ec61ba8>, 'func': <Name.print l.1 at 0x10ec61c18>, 'args': [<Const.str l.1 at 0x10ec61c50>], 'keywords': None}
```

**Comment by li-boxuan** [comment]
The only check other than AST checker that returns `col_offset` is the [_check_space](https://github.com/PyCQA/pylint/blob/master/pylint/checkers/format.py#L731) function. This can be verified by `grep`.

I don't understand why someone adds 1 to col_offset on [line 777](https://github.com/PyCQA/pylint/blob/master/pylint/checkers/format.py#L777). I would argue that it's a bug. To see that, we can run `pylint` on a file that has bad-whitespace issues. Note that this version is 2.0.0. The up-to-date release version is 1.8.4, which would always report 0 column offset for bad white-space issues.
```
C:  1, 3: Exactly one space required after assignment
a =1
  ^ (bad-whitespace)
C:  2, 3: Exactly one space required after assignment
b =  2
  ^ (bad-whitespace)
C:  3, 2: Exactly one space required before assignment
c= 1
 ^ (bad-whitespace)
```
From the previous example, we can see that it uses one-based column offset. I guess that's why the contributor adds one to col_offset. Since this patch is merged recently and violates AST checker's convention, I think we could conclude that `pylint` uses 0-based convention and report this bug to them.

**Comment by li-boxuan** [comment]
Two upstream bugs have been fixed. https://github.com/PyCQA/pylint/issues/2000 https://github.com/PyCQA/pylint/issues/2012

**Comment by jayvdb** [comment]
@li-boxuan , I did a build of coala-bears with coalib, and most of the errors are related to this

You can use b6ed061673 to unblock this.

I have created the following which are also uncovered with this:
- https://github.com/coala/coala-bears/issues/2562
- https://github.com/coala/coala-bears/issues/2563

Could you investigate them all - maybe they can all be fixed with your new tools (https://github.com/coala/coala/issues/5346).  However the linters *may* have varying zero-based or one-based offsets, in which case we need to file upstream bugs quickly and/or possibly need to reduce the error to a warning, so we can use the latest coalib master.
